### PR TITLE
 Laboratory work #1, Bugakova Darya - 20FPL1

### PR DESCRIPTION
**Pull request by daribbb**

### CODE REVIEW

**Comment by ValeriyaKuznetsova** [comment]
It's time to push some changes. October 1st is coming

**Comment by ValeriyaKuznetsova** [comment]
still nothing :(

**Comment by VSvir** [comment]
Added label and requested review as Dasha hasn't got access to the main branch

File: lab_1/main.py [hunk 1]
```diff
@@ -11,7 +11,17 @@ def tokenize(text: str) -> list or None:
     :param text: a text
     :return: a list of lower-cased tokens without punctuation
     """
-    pass
+    symbols = ['!', '@', '#', '$', '%', '^', '&', '*', '(', ')', '-', '_', '=', '+', '[', '{', '}',
+               ']', ';', '<', ',', '>', '.', '?', '|', '\\', ':', ';', '"', "'",
+               '/', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '\t']
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:16]: why don't you like `\t`?

File: lab_1/main.py [hunk 2]
```diff
@@ -11,7 +11,17 @@ def tokenize(text: str) -> list or None:
...
+    if isinstance(text, str):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:17]: put all checks with None at the beginning
like if the argument is incorrect, return None

It'd be easier to read

File: lab_1/main.py [hunk 3]
```diff
@@ -11,7 +11,17 @@ def tokenize(text: str) -> list or None:
...
+        for s in symbols:
+            text = text.replace(s, '')
+        text = text.lower()
+        text = text.split()
+        return text
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:22]: you can return lower split just here

File: lab_1/main.py [hunk 4]
```diff
@@ -21,7 +31,15 @@ def remove_stop_words(tokens: list, stop_words: list) -> list or None:
     :param stop_words: a list of stop words
     :return: a list of tokens without stop words
     """
-    pass
+    if not isinstance(tokens, list) or not isinstance(stop_words, list):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:34]: this is what I was talking about

File: lab_1/main.py [hunk 5]
```diff
@@ -21,7 +31,15 @@ def remove_stop_words(tokens: list, stop_words: list) -> list or None:
...
+        return None
+    checking_tokens = []
+    if isinstance(tokens, list) and isinstance(stop_words, list):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:37]: do you need this check after None checks?

File: lab_1/main.py [hunk 6]
```diff
@@ -30,7 +48,18 @@ def calculate_frequencies(tokens: list) -> dict or None:
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
-    pass
+    freq_dict = {}
+    if not isinstance(tokens, list):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:52]: put it just at the beginning

File: lab_1/main.py [hunk 7]
```diff
@@ -40,7 +69,18 @@ def get_top_n_words(freq_dict: dict, top_n: int) -> list or None:
...
     :return: a list of the most common words
     """
-    pass
+
+    if not isinstance(freq_dict, dict) or not isinstance(top_n, int):
+        return None
+    top_n_words = []
+    freq_dict_keys = [key for key in freq_dict]
+    for i in freq_dict_keys:
+        top_n_words = sorted(freq_dict, key=freq_dict.get, reverse=True)[:top_n]
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:78]: good, but you can sort keys just here without the previous lines
do you really need a loop?

File: lab_1/main.py [hunk 8]
```diff
@@ -51,7 +91,18 @@ def create_language_profile(language: str, text: str, stop_words: list) -> dict
     :param stop_words: a list of stop words
     :return: a dictionary with three keys – name, freq, n_words
     """
-    pass
+
+    if not isinstance(language, str):
+        return None
+    if not isinstance(text, str):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:97]: join checks in one

File: lab_1/main.py [hunk 9]
```diff
@@ -51,7 +91,18 @@ def create_language_profile(language: str, text: str, stop_words: list) -> dict
...
+        return None
+    if not isinstance(stop_words, list):
+        return None
+    tokens = tokenize(text)
+    tokens_without_stop_words = remove_stop_words(tokens, stop_words)
+    frequencies = calculate_frequencies(tokens_without_stop_words)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:103]: transform a text in one line

File: lab_1/main.py [hunk 10]
```diff
@@ -51,7 +91,18 @@ def create_language_profile(language: str, text: str, stop_words: list) -> dict
...
+    language_profile = {'name': language, 'freq': frequencies, 'n_words': len(frequencies)}
+    return language_profile
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:105]: you can return a dict just here

File: lab_1/main.py [hunk 11]
```diff
@@ -74,7 +135,19 @@ def detect_language(unknown_profile: dict, profile_1: dict, profile_2: dict, top
...
+        return None
+    compare_1 = compare_profiles(unknown_profile, profile_1, top_n)
+    compare_2 = compare_profiles(unknown_profile, profile_2, top_n)
+    language = ""
+    if compare_1 > compare_2:
+        language += profile_1['name']
+    elif compare_2 > compare_1:
+        language += profile_2['name']
+    else:
+        language += sorted([profile_1['name'], profile_2['name']])[0]
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:149]: explain what happens here in writing

File: lab_1/main.py [hunk 12]
```diff
@@ -62,7 +113,17 @@ def compare_profiles(unknown_profile: dict, profile_to_compare: dict, top_n: int
...
     :return: the distance
     """
-    pass
+    if not isinstance(unknown_profile, dict) or not isinstance(profile_to_compare, dict) or \
+            not isinstance(top_n, int):
+        return None
+    top_unknown_profile = get_top_n_words(unknown_profile['freq'], top_n)
+    top_profile_to_compare = get_top_n_words(profile_to_compare['freq'], top_n)
+    common_words = []
+    for profile_word in top_profile_to_compare:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:122]: explain what happens in this loop, please

File: lab_1/main.py (hunk 7 above)
Comment (ValeriyaKuznetsova) [reply lab_1/main.py:78]: explain what happens here, please

**Review by ValeriyaKuznetsova — COMMENTED** [review]
Don't forget about `start.py`
You need to classify a text there

In general, it's good. I have some comments on the style and implementation. Also, there are some questions that are obligatory to answer – here or via Telegram.

File: lab_1/main.py (hunk 12 above)
Comment (daribbb) [reply lab_1/main.py:115]: get the tops of the words of two profiles, take each profile_word in top_profile_to_compare, if the profile_word is in top_unknown_profile, add it to the common_words.divide amount of elements in the list of common_words by list of top words and round it.

File: lab_1/main.py (hunk 11 above)
Comment (daribbb) [reply lab_1/main.py:142]: compare compare_1 and compare_2, if compare_1 > compare_2 or compare_1 < compare_2  func returns the appropriate profile name by key, if compare_1 = compare_2 - create a list, sort the names of profiles and take the first element in alphabetical order

Comment (ValeriyaKuznetsova) [reply lab_1/main.py:142]: I don't understand why you need to add a new name to an empty string

**Comment by ValeriyaKuznetsova** [comment]
Be ready to discuss the pending comment today

**Pull request closed without merging**